### PR TITLE
Part 2: connectd onionmessage improvements

### PR DESCRIPTION
**Pull request by rustyrussell**
~(Based on #7454)~ - Merged, rebased!

This enhanced connectd's handling of onion messages: we ratelimit them, we handle forwarding them via scid (not just node_id) and we turn them on by default now they're in the spec.

### CODE REVIEW

File: connectd/onion_message.c [hunk 1]
```diff
@@ -115,17 +115,39 @@ void handle_onion_message(struct daemon *daemon,
...
 	if (!feature_offered(daemon->our_features->bits[NODE_ANNOUNCE_FEATURE],
 			     OPT_ONION_MESSAGES))
 		return;
 
-	/* FIXME: ratelimit! */
+	/* Adjust tokens if they're entitled to more */
+	msec = time_to_msec(timemono_between(now, peer->onionmsg_last_incoming));
+	peer->onionmsg_incoming_tokens += msec;
+	if (peer->onionmsg_incoming_tokens > ONION_MSG_TOKENS_MAX)
+		peer->onionmsg_incoming_tokens = ONION_MSG_TOKENS_MAX;
```
Comment (niftynei) [inline connectd/onion_message.c:130]: would it be worth flipping onionmsg_limit_warned false here? (so resetting after frequency has decreased to less than 4/sec)

**Review by niftynei — APPROVED** [review]
ACK 966b15ebf873eeb62466b8856b53fb6436130cb1

File: lightningd/onion_message.c [hunk 2]
```diff
@@ -249,3 +249,82 @@ static const struct json_command sendonionmessage_command = {
...
+	for (size_t i = 0; i < tal_count(hops); i++)
+		sphinx_add_hop(sphinx_path, &hops[i].node, hops[i].tlv);
+
+	/* BOLT-onion-message #4:
+	 * - SHOULD set `onion_message_packet` `len` to 1366 or 32834.
+	 */
+	if (sphinx_path_payloads_size(sphinx_path) <= ROUTING_INFO_SIZE)
+		onion_size = ROUTING_INFO_SIZE;
+	else
+		onion_size = 32768;
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/onion_message.c:306]: Good catch!  Indeed, 66 bytes is framing overhead.  Added comment!

**Comment by rustyrussell** [comment]
Rebased on master.

File: common/onion_message_parse.c [hunk 3]
```diff
@@ -58,14 +58,27 @@ static bool decrypt_forwarding_onionmsg(const struct pubkey *blinding,
...
+	if (encmsg->next_node_id)
+		sciddir_or_pubkey_from_pubkey(next_node, encmsg->next_node_id);
+	else if (encmsg->short_channel_id) {
+		/* This is actually scid, not sciddir, but the type is convenient! */
+		struct short_channel_id_dir scidd;
+		scidd.scid = *encmsg->short_channel_id;
+		scidd.dir = 0;
+		sciddir_or_pubkey_from_scidd(next_node, &scidd);
+	} else
 		return false;
```
Comment (vincenzopalazzo) [inline common/onion_message_parse.c:81]: This looks like useful information to log, no? like "ignoring message `blabla ..`" but this can be a followup PR

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/7455/commits/e8d244c784b8b1f1fad01f1cb97caceebb1a6806

**Pull request merged**